File: web/frontend/src/services/skillApi.ts
import instance from './request'
import type {
  Skill,
  SkillCreateRequest,
  SkillRepoImportResponse,
  SkillRepoListResponse,
  SkillRepoSource,
} from '@/types/skill'

/** Backend paginated list response */
interface SkillListResponse {
  items: Skill[]
  total: number
  page: number
  page_size: number
}

/** Backend delete response */
interface SkillDeleteResponse {
  deleted_count: number
  failed_count: number
  results: { id: string; name: string; deleted: boolean; reason: string }[]
}

/** Backend upload response */
interface SkillsUploadResponse {
  skills_dir: string
  total: number
  valid_count: number
  invalid_count: number
  skipped_count: number
  results: { name: string; valid: boolean; errors: { field: string; message: string }[] }[]
  skipped: string[]
}

export async function fetchSkills(params?: {
  category?: string
  page?: number
  page_size?: number
  enabled?: boolean
}): Promise<SkillListResponse> {
  const res = await instance.get('/skill/list', { params })
  return res.data.data as SkillListResponse
}

export async function fetchSkill(id: string): Promise<Skill> {
  const res = await instance.get(`/skill/${id}`)
  return res.data.data as Skill
}

export async function createSkill(data: SkillCreateRequest): Promise<Skill> {
  const res = await instance.post('/skill', data)
  return res.data.data as Skill
}

export async function updateSkill(
  id: string,
  data: Partial<SkillCreateRequest & { enabled: boolean }>,
): Promise<Skill> {
  const res = await instance.put(`/skill/${id}`, data)
  return res.data.data as Skill
}

export async function deleteSkill(id: string): Promise<SkillDeleteResponse> {
  const res = await instance.delete(`/skill/${id}`)
  return res.data.data as SkillDeleteResponse
}

export async function deleteSkillsBatch(ids: string[]): Promise<SkillDeleteResponse> {
  const res = await instance.delete('/skill/batch', { data: { ids } })
  return res.data.data as SkillDeleteResponse
}

export async function toggleSkill(id: string, enabled: boolean): Promise<Skill> {
  const res = await instance.patch(`/skill/${id}/toggle`, { enabled })
  return res.data.data as Skill
}

export async function uploadSkills(file: File): Promise<SkillsUploadResponse> {
  const formData = new FormData()
  formData.append('file', file)
  const res = await instance.post('/skill/upload_skills', formData)
  return res.data.data as SkillsUploadResponse
}

export async function searchSkills(
  name: string,
  params?: { page?: number; page_size?: number; category?: string; enabled?: boolean },
): Promise<SkillListResponse> {
  const res = await instance.get('/skill/search', { params: { name, ...params } })
  return res.data.data as SkillListResponse
}

/** 获取后端注册的权威技能仓库来源 */
export async function fetchRepoSkillSources(): Promise<SkillRepoSource[]> {
  const res = await instance.get('/skill/repo/sources')
  return (res.data.data?.sources ?? []) as SkillRepoSource[]
}

/** 抓取指定来源的技能榜单（服务端带缓存与限流保护） */
export async function fetchRepoSkills(params: {
  source: string
  keyword?: string
  page?: number
  page_size?: number
  refresh?: boolean
}): Promise<SkillRepoListResponse> {
  const res = await instance.get('/skill/repo/list', { params })
  return res.data.data as SkillRepoListResponse
}

/** 把榜单中选中的技能导入到 workspace/skills_upload/ 并入库 */
export async function importRepoSkills(
  source: string,
  skillIds: string[],
): Promise<SkillRepoImportResponse> {
  const res = await instance.post('/skill/repo/import', { source, skill_ids: skillIds })
  return res.data.data as SkillRepoImportResponse
}

File: web/backend/src/api/skill/skill_api.py
"""Skill upload API endpoints."""
from fastapi import APIRouter, Depends, File, Query, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from api.deps import get_current_user_id, get_db
from api.skill.schemas import (
    SkillBatchDeleteRequest,
    SkillCreateRequest,
    SkillDeleteResponse,
    SkillInfo,
    SkillListResponse,
    SkillsUploadResponse,
    SkillToggleRequest,
    SkillUpdateRequest,
)
from api.skill.service import (
    create_skill,
    delete_skill,
    delete_skills_batch,
    get_skill,
    list_skills,
    process_skills_upload,
    search_skills,
    toggle_skill_enabled,
    update_skill,
)
from core.decorators import handle_errors
from core.response import ApiResponse, ok

router = APIRouter(prefix="/api/skill", tags=["skill"])


@router.post("/upload_skills", response_model=ApiResponse[SkillsUploadResponse])
@handle_errors
async def upload_skills(
    file: UploadFile = File(...),
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """上传 skills 的压缩包（zip, tar.gz, tar.bz2, tar.xz），校验并安装到 workspace/skills_upload/。

    解压后每个子目录视为一个 skill 包，按 Agent Skills 规范校验 SKILL.md。
    校验失败的 skill 仍会复制到目标目录，在响应中标记 valid=false。
    每个成功安装的 skill 入库到 skills 表。
    """
    result = await process_skills_upload(file, db)
    return ok(result)


@router.get("/list", response_model=ApiResponse[SkillListResponse])
@handle_errors
async def skill_list(
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页条数"),
    category: str | None = Query(None, description="分类筛选"),
    enabled: bool | None = Query(None, description="启用状态筛选"),
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """分页列出所有已入库的技能，按上传时间倒序排列。"""
    result = await list_skills(db, page, page_size, category, enabled)
    return ok(result)


@router.get("/search", response_model=ApiResponse[SkillListResponse])
@handle_errors
async def skill_search(
    name: str = Query(..., min_length=1, description="技能名称关键词"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页条数"),
    category: str | None = Query(None, description="分类筛选"),
    enabled: bool | None = Query(None, description="启用状态筛选"),
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """按技能名称模糊搜索已入库的技能，支持分页、分类和状态筛选。"""
    result = await search_skills(db, name, page, page_size, category, enabled)
    return ok(result)


@router.post("", response_model=ApiResponse[SkillInfo])
@handle_errors
async def skill_create(
    req: SkillCreateRequest,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """手动创建单个技能，写入 SKILL.md 并入库。"""
    result = await create_skill(db, req)
    return ok(result)


@router.get("/{skill_id}", response_model=ApiResponse[SkillInfo])
@handle_errors
async def skill_get(
    skill_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """获取单个技能详情。"""
    result = await get_skill(db, skill_id)
    return ok(result)


@router.put("/{skill_id}", response_model=ApiResponse[SkillInfo])
@handle_errors
async def skill_update(
    skill_id: str,
    req: SkillUpdateRequest,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """更新技能元数据，只更新传入的非空字段。"""
    result = await update_skill(db, skill_id, req)
    return ok(result)


@router.patch("/{skill_id}/toggle", response_model=ApiResponse[SkillInfo])
@handle_errors
async def skill_toggle(
    skill_id: str,
    req: SkillToggleRequest,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """切换技能的启用/禁用状态。"""
    result = await toggle_skill_enabled(db, skill_id, req.enabled)
    return ok(result)


@router.delete("/batch", response_model=ApiResponse[SkillDeleteResponse])
@handle_errors
async def skill_batch_delete(
    req: SkillBatchDeleteRequest,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """批量删除技能，同时删除数据库记录和 workspace 目录。"""
    result = await delete_skills_batch(db, req.ids)
    return ok(result)


@router.delete("/{skill_id}", response_model=ApiResponse[SkillDeleteResponse])
@handle_errors
async def skill_delete(
    skill_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """删除单个技能，同时删除数据库记录和 workspace 目录。"""
    result = await delete_skill(db, skill_id)
    return ok(result)
